### PR TITLE
(PUP-9482) Add an index() function returning index of found value

### PR DESCRIPTION
**Pull request by hlindberg**
This adds the function index() that returns the index of a found value
or `undef` if not found. The function operates on any iterable and
search can be made with explicit value, or a lambda that should return
truthy for the wanted value.

### CODE REVIEW

File: lib/puppet/functions/index.rb [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+# $data = ["routers", "servers", "workstations"]
+# notice $data.index |$value| { $value == 'servers' } # notices 1
+# notice $data.index |$value| { $value == 'hosts'  }  # notices undef
+# ```
+#
+# @example Using the `index` function with a Hash and a one-parameter lambda
+#
+# ```puppet
+# $data = {types => ["routers", "servers", "workstations"], colors => ['red', 'blue', 'green']}
+# notice $data.index |$value| { 'servers' in $value}  # notices 'types'
```
Comment (Magisus) [inline lib/puppet/functions/index.rb]: Missing space after `$value`.

File: spec/unit/functions/index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        file { "$n": ensure => present }
+      MANIFEST
+      expect(catalog.resource(:file, "2")['ensure']).to eq('present')
+    end
+  end
+
+  context "should be callable on Hash with" do
+    it 'a lambda to compute match' do
+      catalog = compile_to_catalog(<<-MANIFEST)
+        $h = {1 => 10, 2 =>20, 3 => 30}
```
Comment (Magisus) [inline spec/unit/functions/index_spec.rb]: Missing space before `20`. Same for the following tests.

File: spec/unit/functions/index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    end
+
+    it 'when using a String and a value' do
+      catalog = compile_to_catalog(<<-MANIFEST)
+        $s = "foobarfuub"
+        $n = $s.index('banana')
+        file { "test": ensure => if $n =~ Undef { present } else {absent} }
+      MANIFEST
+      expect(catalog.resource(:file, "test")['ensure']).to eq('present')
+    end
```
Comment (Magisus) [inline spec/unit/functions/index_spec.rb]: I would kind of like to also see a test for a String and lambda, even though I think under the hood that is kind of covered by the Array test.

Comment (hlindberg) [reply spec/unit/functions/index_spec.rb]: Lines 69-76 has that already ;-)

**Pull request merged**